### PR TITLE
feat: upgrade operator and revamp module

### PR DESCRIPTION
**Pull request by lentidas**
This PR completely revamps this module. It updates the Keycloak operator and the way Keycloak is deployed. It also provides a new way to bootstrap the DevOps Stack realm, since the Keycloak team depreciated the CRDs used before to deploy said resources.

This could be a v1.0.0, as far as I'm concerned.

### CODE REVIEW

**Comment by lentidas** [comment]
I'll mark this PR as draft since for the time being I need to troubleshoot this with @modridi because I could not bootstrap the devops-stack realm using the new CRD.

File: oidc_bootstrap/main.tf [hunk 1]
```diff
@@ -0,0 +1,148 @@
+resource "null_resource" "dependencies" {
+  triggers = var.dependency_ids
+}
+
+resource "keycloak_realm" "devops_stack" {
+  enabled = true
+  realm   = "devops-stack"
+  attributes = {
+    terraform = "true"
+  }
```
[the remark this answers is not in the document]
Comment (lentidas) [reply oidc_bootstrap/main.tf:10]: This is only a tag. This way on Keycloak's WebUI the users can be made aware that the resource were created using Terraform and any changes made manually could be deleted in the future.

File: oidc_bootstrap/main.tf [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    brute_force_detection {
+      permanent_lockout                = false
+      max_login_failures               = 10
+      wait_increment_seconds           = 60
+      quick_login_check_milli_seconds  = 1000
+      minimum_quick_login_wait_seconds = 60
+      max_failure_wait_seconds         = 900
+      failure_reset_time_seconds       = 43200
+    }
+  }
```
[the remark this answers is not in the document]
Comment (lentidas) [reply oidc_bootstrap/main.tf:42]: If you find it too *extreme*, we can delete this. I took it directly from the example available on the documentation for the Keycloak provider.

File: main.tf [hunk 3]
```diff
@@ -112,29 +131,40 @@ resource "argocd_application" "this" {
...
+  ]
 }
 
-resource "random_password" "clientsecret" {
-  length  = 16
-  special = false
+
+resource "null_resource" "wait_for_keycloak" {
+  # Until curl successfully completes the requested transfer, wait 10 seconds and retry for 180 seconds until timeout.
+  # --retry-all-errors makes curl retry even on non-transient HTTP errors (e.g. 404)
```
[the remark this answers is not in the document]
Comment (lentidas) [reply main.tf:142]: You mean the `--retry-all-errors` flag?

File: oidc_bootstrap/main.tf [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+resource "keycloak_realm" "devops_stack" {
+  enabled = true
+  realm   = "devops-stack"
+  attributes = {
+    terraform = "true"
+  }
+
+  display_name      = "DevOps Stack"
+  display_name_html = "<img width='200px' src='https://raw.githubusercontent.com/camptocamp/devops-stack/gh-pages/images/devops-stack-logo_light_by_c2c_black.png' alt='DevOps Stack Logo'/>"
+
```
[the remark this answers is not in the document]
Comment (lentidas) [reply oidc_bootstrap/main.tf:14]: This is the code that adds the DevOps stack logo on the login page. It's a nice to have. Why do you think it's best to delete this?

File: oidc_bootstrap/main.tf [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+resource "keycloak_openid_client_default_scopes" "client_default_scopes" {
+  realm_id  = resource.keycloak_realm.devops_stack.id
+  client_id = resource.keycloak_openid_client.devops_stack.id
+
+  default_scopes = [
+    "profile",
+    "email",
+    "roles",
+    "web-origins",
```
[the remark this answers is not in the document]
Comment (lentidas) [reply oidc_bootstrap/main.tf:95]: Probably not, I have not tested it without. It was taken from the example.

File: oidc_bootstrap/main.tf [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+  first_name     = each.value.first_name
+  last_name      = each.value.last_name
+  email          = each.value.email
+  email_verified = true
+
+  attributes = {
+    terraform = "true"
+  }
+}
```
[the remark this answers is not in the document]
Comment (lentidas) [reply oidc_bootstrap/main.tf:118]: See [comment above](https://github.com/camptocamp/devops-stack-module-keycloak/pull/6#discussion_r1111706990), this is only a tag to let users on the web interface know that the resource is terraformed.

File: main.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (lentidas) [reply main.tf]: The flag `--retry-all-errors` was added on version 7.71.0 of `curl`. I've tested on my machine and I cannot make the command work without that flag as `curl` will not retry if it receives a 404. In order to retry on that case you will need the `--retry-all-errors` and the `-f` flag in combination, from what I understand from the official documentation.

In conclusion, this poses a problem when a user has an outdated version of `curl` on their machine. In my opinion we should support the more recent versions unless you have a workaround.

**Review by modridi — DISMISSED** [review]

**Review by modridi — DISMISSED** [review]

**Review by modridi — APPROVED** [review]

**Pull request merged**